Add tutorial on using Crossplane AWS provider with LocalStack (#804)
localstack/docs: content/en/user-guide/integrations/crossplane/index.md

---
title: "Crossplane"

> tags: ["crossplane", "kubernetes", "infrastructure-as-code"]

Reviewer: Does it help SEO to put "localstack" in the tags?

weight: 5
description: >
  Use the Crossplane cloud-native control plane framework with LocalStack
aliases:

>   - /integrations/crossplane/

Reviewer 2: There is no need to put this :)

---

<img src="logo-crossplane.svg" width="500px" alt="Crossplane logo">

## Overview

[Crossplane](https://www.crossplane.io) is a cloud-native control plane framework, which offers an extensible backend that enables orchestrating applications and infrastructure via declarative APIs and resource definitions.

Crossplane offers a native [AWS provider](https://github.com/upbound/provider-aws) which can be used to create and manage AWS cloud resources via the Crossplane platform.
For example, it can be used to create S3 buckets, SQS queues, Lambda functions, among many other resources.

> At the time of writing, the Crossplane AWS provider supports a comprehensive set of some [900+ resource types](https://marketplace.upbound.io/providers/upbound/provider-aws).

Reviewer 2: ```suggestion
Crossplane AWS provider supports a comprehensive set of some [900+ resource types](https://marketplace.upbound.io/providers/upbound/provider-aws).
```

> ## Tutorial: Creating AWS resources in LocalStack via Crossplane

Reviewer 2: ```suggestion
## Getting started
```

In the following, we provide a step-by-step guide for installing Crossplane in a local test environment, and creating AWS resources (S3 bucket, SQS queue) in LocalStack via Crossplane.

> ### Step 1: Installing Crossplane in local Kubernetes
> 
> As a prerequisite, we need the following:

Reviewer 2: ```suggestion
### Prerequisites
```

* LocalStack running in local Docker
* A local Kubernetes cluster:
  * We can use the [embedded Kubernetes cluster](https://docs.docker.com/desktop/kubernetes) that ships with modern versions of Docker Desktop (can be easily enabled in the Docker settings)
  * Alternatively, you can [create a local EKS cluster](https://docs.localstack.cloud/user-guide/aws/elastic-kubernetes-service/#create-an-embedded-kubernetes-cluster) in LocalStack directly, which will spin up a light-weight embedded `k3d` Kubernetes cluster in your Docker environment
* The [`helm`](https://helm.sh) and [`kubectl`](https://kubernetes.io/docs/tasks/tools/#kubectl) command-line clients installed

> Once your `kubectl` is configured to point to the local Kubernetes cluster, we first install Crossplane via `helm`:

Reviewer 2: ```suggestion
## Installing Crossplane in local Kubernetes

Once your `kubectl` is configured to point to the local Kubernetes cluster, we first install Crossplane via `helm`:
```

{{<command>}}
$ helm repo add crossplane-stable https://charts.crossplane.io/stable
$ helm repo update
$ helm install crossplane crossplane-stable/crossplane --namespace crossplane-system --create-namespace
{{</command>}}

The installation may take a few minutes. In parallel, we can install the `crossplane` command-line extensions for `kubectl`:

> {{<command>}}

Reviewer: Do you want to separate these two commands and remove the `...`?

$ curl -sL https://raw.githubusercontent.com/crossplane/crossplane/master/install.sh | bash
...
$ sudo mv kubectl-crossplane /usr/local/bin
{{</command>}}

To confirm that the installation was successful, we can run these `kubectl` commands, which should yield output similar to the following:
{{<command>}}
$ kubectl crossplane --version

> v1.13.2

Reviewer: I think we want to remove output from command sections so copy and paste works better?

Author: Good point - see my comment in the PR description above. I'm sort of a fan of the terse syntax with commands&output in a single listing, as otherwise the length of the text kind of explodes. Will change it if it's the general consensus, but I'm not sure it makes for a great visual experience if we _always_ have a separate listing with the command, followed by some prose text like _"This command will yield the following output"_ , followed by another listing with the output text. 

What I would basically like to have is a `<command>` tag that has two properties:
1. special handling of `$` (not getting selected when user selects the code with the mouse)
2. **no** "Copy" button in the top right corner

If I use a standard code block with backticks, it doesn't satisfy point (1), and if I use a `<command>` tag, it doesn't satisfy point (2). 🤷 

Maybe @HarshCasper or @lukqw @webdev51 can help us out here..?

Reviewer 3: Hi, I made 2 improvements to the `command.html` file which is responsible for rendering the `<command>` short code .

1. Added an option to pass a prop called `wrapper` which (as the name suggests) will simply wrap the `command` box in a wrapper `div` which won't do anything but the thing is, if you pass the wrapper called `wrapper="disable-copy-btn"`, we have some styles going on which will hide the copy button. So, it'll look like this:  `<command wrapper="disable-copy-btn">}} ... {{</command>` and we'll get a command with no copy button

2. I also added a wrapper tag called  `<disable-copy>` which works like this:
Let's say we've a command with some output:
`{{<command>}}
$ kubectl crossplane --version
<disable-copy>v1.13.2</disable-copy>
{{</command>}}`

The above command will still have a copy button but clicking "Copy"  will exclude that text inside the `disable-copy` tags

Please let me know if anything can be improved in this and can I push the changes in the same PR? Both the examples are there in my local branch for now. I can also create a new PR if that's better. 

CC: @lukqw , @whummer , @HarshCasper

Author: Hey @webdev51 , thanks for jumping on this so quickly! 🚀 

I think both options are great. 👌  I have a slight preference for `<disable-copy>`, as it combines the best of both worlds - allowing the copying of the command, while still allowing non-copyable output in the code listing. 

Please feel free to push your changes to the branch directly! 🙌

Reviewer 3: @whummer, I did the changes with `<disable-copy>` but still left the `wrapper` functionality as well for future if we ever need to completely hide the "Copy" button.

Would be great if you just have a quick final look :smile: 

I'll also mention it in the #sig-docs maybe so everyone is aware of these changes

$ kubectl get crds | grep crossplane
compositions.apiextensions.crossplane.io                     2023-09-03T11:30:36Z
configurations.pkg.crossplane.io                             2023-09-03T11:30:36Z
...
{{</command>}}

> ### Step 2: Installing the Crossplane AWS Provider

Reviewer 2: ```suggestion
Installing the Crossplane AWS Provider
```

Author: Should we highlight these section titles, at least with bold font, or a `###` prefix?

Once the basic Crossplane installation is running properly, we can proceed with installing the AWS provider.
Newer versions of Crossplane promote the use of [provider families](https://docs.upbound.io/providers/provider-families), which are collections of providers for different groups of resources.
For example, there is a separate provider for each individual AWS service (like S3, SQS, Lambda, etc), and in addition provider family provides shared resources for common configuration of all services (e.g., credentials, etc).

In the following, we first install the AWS provider for S3.

> Please note that you can copy/paste the entire multi-line command below into your terminal:

Reviewer 2: ```suggestion
Note that you can copy/paste the entire multi-line command below into your terminal:
```

We should try to avoid words like "please" in technical docs.

{{<command>}}
$ cat <<EOF | kubectl apply -f -
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: provider-aws-s3
spec:
  package: xpkg.upbound.io/upbound/provider-aws-s3:v0.40.0
EOF
{{</command>}}

We also install the AWS provider for SQS:
{{<command>}}
$ cat <<EOF | kubectl apply -f -
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: provider-aws-sqs
spec:
  package: xpkg.upbound.io/upbound/provider-aws-sqs:v0.40.0
EOF
{{</command>}}

After some time, the providers should get into healthy state, which can be confirmed via `kubectl get providers`:
{{<command>}}
$ kubectl get providers

> NAME                          INSTALLED   HEALTHY   PACKAGE                                               AGE

Reviewer: Move output to non-command section so copy-paste works better?

Author: Will be handled by the changes from @webdev51 mentioned above.. 👍 (same for other comments in this PR)

stringData:
  creds: |
    [default]
    aws_access_key_id = test
    aws_secret_access_key = test
EOF
{{</command>}}

Finally, we create an AWS  `ProviderConfig` that references the secret created above, and defines a static `endpoint` pointing to the LocalStack URL `http://host.docker.internal:4566`:
{{<command>}}
$ cat <<EOF | kubectl apply -f -
apiVersion: aws.upbound.io/v1beta1
kind: ProviderConfig
metadata:
  name: default
spec:
  credentials:
    source: Secret
    secretRef:
      name: localstack-aws-secret
      namespace: default
      key: creds
  endpoint:
    hostnameImmutable: true
    # TODO: add more services to this list, as needed
    services: [iam, s3, sqs, sts]
    url:
      type: Static
      static: http://host.docker.internal:4566
  skip_credentials_validation: true
  skip_metadata_api_check: true
  skip_requesting_account_id: true
  s3_use_path_style: true
EOF
{{</command>}}

{{<alert title="Note">}}
The endpoint `http://host.docker.internal:4566` in the listing above assumes that you are running Kubernetes in the local Docker engine, and that LocalStack is up and running and available on default port `4566`.
{{</alert>}}

{{<alert title="Note">}}

> The Crossplane AWS provider currently requires us to specify the list of `services` for which the local `endpoint` is used as the target URL. Please make sure to extend this list accordingly if you're working with additional LocalStack services.

Reviewer: consider adding
> The YAML above must configure the list of `services` since the Crossplane AWS provider requires it. 
Or something like that. 
I just read the `alert` and I was like "uh where is this referring to".

{{</alert>}}

> ### Step 3: Deploying sample resources in LocalStack

Reviewer 2: ```suggestion
Deploying sample resources in LocalStack
```

After the Crossplane AWS provider is properly installed and configured, we can proceed with creating some local resources.

First, we create an S3 bucket named `crossplane-test-bucket`:
{{<command>}}
$ cat <<EOF | kubectl apply -f -
apiVersion: s3.aws.upbound.io/v1beta1
kind: Bucket
metadata:
  name: crossplane-test-bucket
spec:
  forProvider:
    region: us-east-1
EOF
{{</command>}}

If everything is wired up correctly, you should now see some activity in the LocalStack log outputs, where Crossplane starts deploying the S3 bucket against LocalStack.
After some time, the bucket should be transitioning into `ready` state within Crossplane:
{{<command>}}
$ kubectl get buckets

> NAME                     READY   SYNCED   EXTERNAL-NAME            AGE

Reviewer: More output in command section

crossplane-test-bucket   True    True     crossplane-test-bucket   30s
{{</command>}}

... and the bucket it should also be visible when querying the local S3 buckets in LocalStack via [`awslocal`](https://github.com/localstack/awscli-local):
{{<command>}}
$ awslocal s3 ls

> 2023-09-03 15:18:47 crossplane-test-bucket

Reviewer: Output in command section

{{</command>}}

We can repeat the same exercise for creating a local SQS queue named `crossplane-test-queue`:
{{<command>}}
$ cat <<EOF | kubectl apply -f -
apiVersion: sqs.aws.upbound.io/v1beta1
kind: Queue
metadata:
  name: crossplane-test-queue
spec:
  forProvider:
    name: crossplane-test-queue
    region: us-east-1
EOF
{{</command>}}

After some time, the queue should transition into `ready` state in Crossplane:
{{<command>}}
$ kubectl get queues

> NAME                    READY   SYNCED   EXTERNAL-NAME                                                         AGE

Reviewer: Output in command section

crossplane-test-queue   True    True     http://host.docker.internal:4566/000000000000/crossplane-test-queue   40s
{{</command>}}

... and the queue should be visible when listing the SQS queues in LocalStack:
{{<command>}}
$ awslocal sqs list-queues

> {

Reviewer: Output in command section

    "QueueUrls": [
        "http://localhost:4566/000000000000/crossplane-test-queue"
    ]
}
{{</command>}}

### Summary

The Crossplane AWS provider is a great way to manage AWS resources, and by leveraging the `endpoint` configuration of the provider, we can seamlessly run resource deployments against LocalStack.

In this tutorial, we have provided an end-to-end walkthrough of how to provision two simple resources - an S3 bucket, and an SQS queue. Crossplane supports a vast range of additional AWS resource types, as well as advanced operations like updating, deleting, or composing resources.

> Please refer to the additional reading material to learn and explore more advanced features.

Reviewer 2: ```suggestion
You can refer to the additional reading material to learn and explore more advanced features.
```
